google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java

```diff
@@ -33,6 +95,334 @@
    */
   @Nullable
   CloudProject showDialog(@Nullable CloudProject cloudProject) {
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
```

Comment:
I'm trying to wrap my head around this. Why didn't you want this class to be a `DialogWrapper` itself?

Reply:
in other words, why not just extend from DialogWrapper? As is it is, it is very tightly coupled to DialogWrapper anyway, so I'm not seeing the advantage. I could be wrong though.

Reply:
I tried it first and inheriting from `DialogWrapper` brings a lot of internal UI implementation with it - I had to spy/mock/override a lot of internal methods, and it started 'leaking projects' because I used it - something was not disposing properly in mock test mode. It's much easier and cleaner to test and edit as a simple form, but we have to have this wrapper class to pass dialog wrapper buttons and additional actions. It is not very friendly to composition instead of inheritance.